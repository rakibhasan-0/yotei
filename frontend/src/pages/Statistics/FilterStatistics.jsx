import React, { useContext } from "react"
import FilterContainer from "../../components/Common/Filter/FilterContainer/FilterContainer"
import CheckBox from "../../components/Common/CheckBox/CheckBox"
import DatePicker from "../../components/Common/DatePicker/DatePicker"
import BeltPicker from "../../components/Common/BeltPicker/BeltPicker"
import style from "./FilterStatistics.module.css"


/**
 * 
 * The work is on progress for the statistics page.
 */

export default function FilterStatistics({ onToggleExercise, onToggleKihon, onDateChanges, onToggleBelts, onClearBelts, belts}) {
  return (
    <div>
      <FilterContainer id="filter-container" title="Filtering" numFilters={0}>
        <div className={style.dateContainer}>
          <h2>Från</h2>
          <div></div>
          <DatePicker
            id="start-date-picker"
            //selectedDate={date.startDate}
			//minDate={}
            onChange={(e) => {
              onDateChanges("from", e.target.value);
            }}
          />
        </div>

        <div className={style.dateContainer}>
          <h2>Till</h2>
          <div></div>
          <DatePicker
            id="end-date-picker"
            //selectedDate={getFormattedDateString(new Date())}
            onChange={(e) => {
              onDateChanges("to", e.target.value);
            }}
          />
        </div>

        <BeltPicker
          id={"techniqueFilter-BeltPicker"}
          onToggle={onToggleBelts}
          states={belts}
          onClearBelts={onClearBelts}
          filterWhiteBelt={false}
        />

        <div className={style.checkboxContainer}>
          <h2>Visa Övningar</h2>
          <div></div>
          <CheckBox
            id={"techniqueFilter-VisaÖvningar"}
            checked={false}
            onClick={(isChecked) => {
              console.log("Checkbox exercise clicked", isChecked);
              onToggleExercise(isChecked);
            }}
          />
        </div>

        <div className={style.checkboxContainer}>
          <h2>Kihon</h2>
          <div></div>
          <CheckBox
            id={"techniqueFilter-KihonCheck"}
            checked={false}
            onClick={(isChecked) => {
              console.log("Checkbox kihon clicked", isChecked);
              onToggleKihon(isChecked);
            }}
          />
        </div>
      </FilterContainer>
    </div>
  );
}
