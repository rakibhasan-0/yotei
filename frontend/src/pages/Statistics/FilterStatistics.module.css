.dateContainer {
	display: flex;
	text-align: center;
	margin-bottom: 4%;
}

.checkboxContainer {
	display: flex;
	justify-content: center;
	margin-bottom: 2%;
}

.dateContainer h2 {
	margin-top: 4%;
}

.dateContainer>div {
	width: 75%;
}


.checkboxContainer>h2 {
	margin-right: 10%;
	white-space: nowrap;
}

.checkboxContainer>div {
	width: 100%;
}