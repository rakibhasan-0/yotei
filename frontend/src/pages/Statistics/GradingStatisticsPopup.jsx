import React, { useEffect, useState } from "react"
import Popup from "../../components/Common/Popup/Popup"
import style from "./GradingStatisticsPopup.module.css"
import Dropdown from "../../components/Common/List/Dropdown"
import GradingProtocolsRowsMenu from "../../components/Grading/GradingProtocolsRowsMenu"
import GradingProtocolsRows from "../../components/Grading/GradingProtocolsRows"
import Spinner from "../../components/Common/Spinner/Spinner"

/**
 *  The component for the grading statistics popup. It shows what techniques are required for each belt.
 *  And what techniques have been completed by the group.
 * 
 * 
 * @param {String} id - The id of the component, for testing purposes
 * @param {String} groupID - The id of the group
 * @param {Array} belts - The belts of the group
 * 
 * @since 2024-05-20
 * @author Team Coconut
 */


export default function GradingStatisticsPopup({ id, groupID, belts,datesFrom,datesTo}) {

	const [showPopup, setShowPopup] = useState(false)
	const [protocols, setProtocols] = useState([])
	const [chosenProtocol, setChosenProtocol] = useState("")
	const [beltID, setBeltID] = useState(null)
	const [loading, setLoading] = useState(false)
	const [data, setData] = useState([])

	useEffect(() => {
		if (belts.length > 0) {
			setProtocols(belts.map(belt => belt.name))
			setChosenProtocol(belts[0].name)
			setBeltID(belts[0].id)
		}
	}, [belts])

	useEffect(() => {
		if (groupID && beltID !== null) {
			const fetchGroupGradingProtocol = async () => {
				try {
					setLoading(true)
					const response = await fetch(`/api/statistics/${groupID}/grading_protocol?beltId=${beltID}&startdate=${datesFrom}&enddate=${datesTo}`)
					if (!response.ok) {
						throw new Error("Failed to fetch group data")
					}
					if(response.status === 204){
						//For belts that do not have grading protocols
						setProtocols([])
					} else {
						const groups = await response.json()

						setData(groups)

					}
				} catch (error) {
					console.error("Fetching error:", error)
				} finally {
					setLoading(false)
				}
			}
			fetchGroupGradingProtocol()
		}
	}, [groupID, beltID,showPopup])

	useEffect(() => {
		if (chosenProtocol) {
			const selectedBelt = belts.find(belt => belt.name === chosenProtocol)
			if (selectedBelt) {
				setBeltID(selectedBelt.id)
			}
		}
	}, [chosenProtocol])

	const togglePopup = () => {
		setShowPopup(!showPopup)
	}

	const onSelectRow = (protocol) => {
		setChosenProtocol(protocol)
	}

	return (
		<div className={style.gradingStatisticsContainer} id={id}>
			<div className={style.gradingButtonContainer}>
				<button onClick={togglePopup} id="popup-button">
					<img src="/GradingStatistics.svg" alt="Grading Statistics" />
				</button>
			</div>

			<Popup
				title={"Graderingsprotokoll"}
				setIsOpen={setShowPopup}
				isOpen={showPopup}
				id="grading-statistics-popup"
			>
				<Dropdown text={data.code ? data.code + " " + data.name : "VÃ¤lj ett protokoll"} id="grading-protocols-dropdown">
					<GradingProtocolsRowsMenu protocols={protocols} onSelectRow={onSelectRow} />
				</Dropdown>
				{loading ? <Spinner /> : 
					(data.belt && data.categories) && 
					<GradingProtocolsRows data={data.categories} beltColors={[data.belt]} />
				}
			</Popup>
		</div>
	)
}
