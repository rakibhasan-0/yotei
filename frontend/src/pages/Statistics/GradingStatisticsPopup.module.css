.gradingButtonContainer button {
    border: none;
    background: none;
    position: absolute;
    top: 0;
    right: 3.2em;
}

.gradingButtonContainer img{
    height: 2.5em;
}