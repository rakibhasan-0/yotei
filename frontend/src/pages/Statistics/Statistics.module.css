
.buttonContainer{
	margin-top: 5%;
	margin-bottom: 3%;
}

.FilterAndSortContainer {
	position: relative;
	margin-top: 4%;
}

.activitiesTextContainer{
	display: flex;
	justify-content: space-between;
	align-items: center;
	border-bottom: 1px solid gray;
	margin-top: 3%;
	padding: 0 0 5px;
}


.activitiesTextContainer h5{
	font-size: 20px;

}

.activitiesContainer h5{
	font-size: 20px;
	left: 40px;
}