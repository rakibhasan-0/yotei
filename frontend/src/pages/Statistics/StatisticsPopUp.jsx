import React, { useState,useEffect } from "react"
import style from "./StatisticsButton.module.css"
import Popup from "../../components/Common/Popup/Popup"
import StarRatings from "react-star-ratings"
import BeltColorChart from "../../components/Common/BeltColorChart/BeltColorChart"

/**
	 * Statistics pop-up component.
	 * @input groupActivities - The activities to be displayed in the pop-up.
	 * @input dates - The date interval of the sessions
	 * @input averageRating - The average rating of the sessions
	 * @input numberOfSessions - The number of sessions done in the interval
	 * @returns A pop-up with statistics.
	 * 
	 * @author Team Coconut (Gabriel Morberg)
	 * @since 2024-05-08
	 * @version 1.0
	 */
export default function StatisticsPopUp({groupActivities,dates,averageRating,numberOfSessions}) {
	const startDate = dates.from
	const endDate = dates.to
	const [beltColorsData, setBeltColorsData] = useState({})

	useEffect(() => {
		// Function to calculate the amount of techniques by belt color
		const calculateBeltColorsData = () => {
			const beltColors = {}
			groupActivities.forEach(activity => {
				
				if(activity.type == "technique") {
					activity.beltColors.forEach(beltColor => {
						const identifier = beltColor.belt_name + (beltColor.is_child ? "_c" : "")
						if (beltColors[identifier]) {
							beltColors[identifier]["count"] += activity.count
						} else {
							beltColors[identifier] = {"count":activity.count,
								"color":beltColor.belt_color,
								"isChild":beltColor.is_child}
						}
					})
				}
			})

			const entries = Object.entries(beltColors)
			entries.sort((a, b) => b[1].count - a[1].count)
			const sortedBeltColors = Object.fromEntries(entries)
			setBeltColorsData(sortedBeltColors)
		}
		
		// Call the function to calculate belt colors data when groupActivities change
		calculateBeltColorsData()
	}, [groupActivities])


	const [showPopup, setShowPopup] = useState(false)

	const togglePopup = () => {
		setShowPopup(!showPopup)

	}

	return (
		<div className={style.statisticsButtonContainer}>
		<button onClick={togglePopup}>
			<img src="/Statistics.svg" />
		</button>

			<Popup
			title={"Sammanställning av tillfällen"}
			id="statistics-popup"
			isOpen={showPopup}
			setIsOpen={setShowPopup}
			>
				<div className={style.statisticsPopupContainer}>
					<StarRatings
						rating={averageRating}
						starRatedColor="#ffcc00"
						numberOfStars={5}
						name="rating"
						starDimension="50px"
						starSpacing="4px"
					/>
					<p style={{ color: "#b9b0b0" }}>
						Genomsnittligt betyg {averageRating}/5 för {numberOfSessions}{" "}
						tillfällen från {<br />} datum {startDate} till {endDate}
					</p>
					<p style={{ fontSize: "25px" }}>Bält-tekniker</p>
					<div style={{ overflowY: "scroll" }}>
						<BeltColorChart beltColorsData={beltColorsData} />
					</div>
				</div>
			</Popup>
		</div>
	);
}

