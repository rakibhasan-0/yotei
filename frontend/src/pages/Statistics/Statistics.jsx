import React, { useState, useEffect, useContext } from "react"
import { useParams, useNavigate } from "react-router-dom"
import { AccountContext } from "../../context"
import Spinner from "../../components/Common/Spinner/Spinner"
import style from "./Statistics.module.css"
import Button from "../../components/Common/Button/Button"
import TechniqueCard from "../../components/Common/Technique/TechniqueCard/TechniqueCard"
import StatisticsPopUp from "./StatisticsPopUp"
import FilterStatistics from "./FilterStatistics"
import {getFormattedDateString} from "../../components/Common/DatePicker/DatePicker"

/**
 * 
 *  The work is on progress for the statistics page.
 * 
 */

export default function Statistics() {

	const navigate = useNavigate()
	const { groupID } = useParams()
	const [groupName, setGroupName] = useState(null)
	const [loading, setLoading] = useState(true)
	const { token } = useContext(AccountContext)
	const [groupActivities, setGroupActivities] = useState([])
	const [selectedBelts, setSelectedBelts] = useState([])
	const [filter, setFilter] = useState({
		showExercises: false,
		showKihon: false,
	})


	// creating a date object for two years before from now and today's date
	const twoYearsBeforeFromNow = new Date()
	twoYearsBeforeFromNow.setFullYear(twoYearsBeforeFromNow.getFullYear() - 2)
	const today = new Date()

	// state for storing the dates
	const [dates, setDates] = useState({
		from: getFormattedDateString(twoYearsBeforeFromNow),
		to: getFormattedDateString(today),
	})

	// filtering the group activities based on the selected belts. 
	const activities =	
	selectedBelts.length > 0	
		? groupActivities.filter((activity) =>
			activity.beltColors?.some((belt) =>	
				selectedBelts.some((selectedBelt) =>	
					selectedBelt.child	
						? belt.is_child == true && selectedBelt.name === belt.belt_name
						:	selectedBelt.name === belt.belt_name && belt.is_child == false
				)
			)
		)
		: groupActivities



	function handleBeltToggle(isSelected, belt) {
		setSelectedBelts(prevSelected => {
			if (isSelected) {
				return [...prevSelected, belt]
			} else {
				return prevSelected.filter(b => b.id !== belt.id)
			}
		})
	}


	function onBeltsClear() {
		setSelectedBelts([])
	}
	

	function checkIfDateIsValid(date) {
		return /^[12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/.test(date) && !isNaN(new Date(date).getTime())
	}

	useEffect(() => {

		async function fetchGroupActivitiesData() {	
			if(!checkIfDateIsValid(dates.from) || !checkIfDateIsValid(dates.to)) {
				return
			}
			
			const param = new URLSearchParams({
				kihon: filter.showKihon ? "true" : "false",
				showexercises: filter.showExercises ? "true" : "false",
				startdate: dates.from ? dates.from : "",
				enddate: dates.to ? dates.to : ""
			})

			
		
			try {
				setLoading(true)
				const responseFromGroupNameAPI= await fetch("/api/plan/all", { headers: { token } })
				const responseFromGroupDetailsAPI = await fetch(`/api/statistics/${groupID}?${param}`, {headers: { token }})

				if(responseFromGroupDetailsAPI.status === 200) {
					const data = await responseFromGroupDetailsAPI.json()
					setGroupActivities(data.activities)
				}

				if (!responseFromGroupDetailsAPI.ok || !responseFromGroupNameAPI.ok) {
					throw new Error("Failed to fetch group data")
				}
				
				const groups = await responseFromGroupNameAPI.json()	
				const name = groups.find((group) => group.id === parseInt(groupID))
				setGroupName(name)
				
			}
			catch (error) {
				console.error("Fetching error:", error) // proper error handling will be added later
			}
			finally {
				setLoading(false)
			}
		}
	
		fetchGroupActivitiesData()

	}, [groupID, token, filter, dates])


	function handleDateChanges(variableName, value) {

		const newDate = new Date(value)	
		const endDate = new Date(dates.to)
		
		if(variableName === "from"){
			setDates(newDate > endDate ? {from: dates.from, to: dates.to} : {...dates, from: value})
		}

		else if(variableName === "to") {
			setDates({ ...dates, [variableName]: value })			
		}

	}


	function handleChanges(variableName, value) {
		setFilter({ ...filter, [variableName]: value })
	}

	return (
		<div>
			<title>Statistik</title>
			{loading ? (
				<Spinner />
			) : (
				<h1 style={{ fontSize: "35px" }} id = "statistics-header" >
					{groupName ? `${groupName.name}` : "Gruppen hittades inte"}
				</h1>
			)}

			<div className={style.FilterAndSortContainer}>
				<FilterStatistics
					id="statistics-filter-container"
					onToggleExercise={(value) => handleChanges("showExercises", value)}
					onToggleKihon={(value) => handleChanges("showKihon", value)}
					onDateChanges={handleDateChanges}
					onToggleBelts={handleBeltToggle}
					onClearBelts={onBeltsClear}
					belts={selectedBelts}
					dates={dates}
				/>

				<StatisticsPopUp groupActivities ={groupActivities} filteredActivities = {activities} dates ={dates} />
			</div>

			<div className="activitiesContainer" id="technique-exercise-list">
				{	activities.length === 0 ? <h5 style={{ fontSize: "25px" }}>Inga aktiviteter hittades</h5> :
					activities.map((activity, index) => (
						<TechniqueCard key={index} technique={activity} checkBox={false} id={activity.activity_id} />
					))
				}
			</div>

			<div className={style.buttonContainer}>
				<Button width="25%" outlined={true} onClick={() => navigate(-1) } id ="statistics-back-button">
					<p>Tillbaka</p>
				</Button>
			</div>
		</div>
	)
}	
