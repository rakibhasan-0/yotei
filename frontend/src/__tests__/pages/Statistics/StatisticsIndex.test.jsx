import "@testing-library/jest-dom"
import {render, screen, fireEvent, configure, waitFor} from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"
import Statistics from "../../../pages/Statistics/StatisticsIndex"
import StatisticsPopUp from "../../../pages/Statistics/StatisticsPopUp"
import FilterStatistics from "../../../pages/Statistics/FilterStatistics"
import GradingStatisticsPopup from "../../../pages/Statistics/GradingStatisticsPopup"
import React from "react";
configure({ testIdAttribute: "id" })

/**
 * @author Team Coconut
 * @since 2024-05-02
 * @version 1.0
 */
const mockedGroup = [{
	id: 1,
	name: "Sample Group",
	color: "0C7D2B",
	child: false
}]

const mockedBelts = [{
	id: 5,
	name: "Svart",
	color: "000000",
	child: false
}, {
	id: 2,
	name: "Vitt",
	color: "FFFFFF",
	child: false
}, {
	id: 3,
	name: "Grönt",
	color: "0C7D2B",
	child: false
}, {
	id: 4,
	name: "Brunt",
	color: "83530C",
	child: false
},]




const mockedGroupActivities = [
	{
		activity_id: 1,
		beltColors: [{
			belt_color: "0C7D2B",
			belt_name: "Grönt",
			is_child: false
		}],
		count: 6,
		name: "One punch",
		type: "technique",
	}, {
		activity_id: 2,
		beltColors: [{
			belt_color: "0C7D2B",
			belt_name: "Grönt",
			is_child: false
		}],
		count: 3,
		name: "Two punch",
		type: "technique",
	}, {
		activity_id: 3,
		beltColors: [{
			belt_color: "83530C",
			belt_name: "Brunt",
			is_child: false
		}],
		count: 1,
		name: "Hook mot lever",
		type: "technique",
	}]

describe("Grading popup", () => {

	const popupBtnID = "popup-button"
	const headerText = "Graderingsprotokoll"
	const dropdownText = "Välj ett protokoll"
	const dropdownID = "grading-protocols-dropdown-dropdown"
	const protocolExist = "SVART BÄLTE"
	const protocolNotExist = "BRUNT BÄLTE"
	const protocols= [
		{
			name: "SVART",
			id: 1
		}
	]

	test("Render button", () => {

		// render popup
		render(
			<BrowserRouter>
				<Statistics/>
			</BrowserRouter>
		)

		// statistics page should include the grading popup feature
		expect(screen.getByTestId("grading-statistics-container")).toBeInTheDocument()
		expect(screen.getByTestId(popupBtnID)).toBeInTheDocument()
	})

	test("Access popup", async () => {

		// render page
		render(<BrowserRouter> <Statistics/> </BrowserRouter>)

		// popup elements should not be present before toggling popup
		expect(screen.queryByRole("heading", { name: headerText })).not.toBeInTheDocument()
		expect(screen.queryByText(dropdownText)).not.toBeInTheDocument()

		// toggle popup
		fireEvent.click(screen.getByTestId(popupBtnID))

		// some popup elements should be visible on toggle
		expect(screen.getByRole("heading", { name: headerText })).toBeInTheDocument()
		expect(screen.getByText(dropdownText)).toBeInTheDocument()
		
	})

	test("Functionality", async () => {

		// render popup
		render(<BrowserRouter>
			<GradingStatisticsPopup id = {"grading-statistics-container"} groupid = {"3"} belts = {protocols}/>
		</BrowserRouter>)

		// toggle popup & dropdown
		fireEvent.click(screen.getByTestId(popupBtnID))
		fireEvent.click(screen.getByTestId(dropdownID))

		// protocols should now be visible
		expect(screen.getByText(protocolExist)).toBeInTheDocument()
		expect(screen.queryByText(protocolNotExist)).not.toBeInTheDocument()
	})
})



/**
 * @author Team Coconut
 * @since 2024-05-02
 * @version 1.0
 */
describe("Statistics Popup", () => {
	test("Clicking button should show popup", async () => {
		// Mock data
		const mockData = {
			groupActivities: mockedGroupActivities,
			dates: {
				from: new Date("2022-05-08").toISOString(),
				to: new Date("2024-05-08").toISOString(),
			},
			averageRating: 4,
			numberOfSessions: 10,
		}

		// Mock the fetch function
		// eslint-disable-next-line no-undef
		global.fetch = jest.fn(() =>
			Promise.resolve({
				ok: true,
				json: () => Promise.resolve(mockData),
			})
		)

		// Render the StatisticsPopup component with mock data
		render(
			<BrowserRouter> 
				<StatisticsPopUp 
					groupActivities={mockData.groupActivities} 
					dates={mockData.dates} 
					averageRating={mockData.averageRating} 
					numberOfSessions={mockData.numberOfSessions} />
			</BrowserRouter>
		)

		// Simulate a click on the button
		fireEvent.click(screen.getByRole("button"))

		// Wait for the popup to appear

		await expect(screen.getByText("Sammanställning av tillfällen")).toBeInTheDocument()
		await expect(screen.getByText("Bält-tekniker")).toBeInTheDocument()

	})
})


describe("Statistics component", () => {

	// Test is incomplete, see bottom of test for in-depth description
	test("renders groups statistic page when data is fetched successfully", async () => {

		// Mock fetch API to return sample data
		// eslint-disable-next-line no-undef
		global.fetch = jest.fn(() =>
			Promise.resolve({
				ok: true,
				json: () =>
					Promise.resolve([
						...mockedGroupActivities,
						...mockedGroup
					]),
			})
		)

		// Render Statistics page
		render(<BrowserRouter> <Statistics /> </BrowserRouter>)

		// Wait for groups statistic page to be rendered
		await screen.findByText("Sample Group")

		// Ensure "Sample Group" text is rendered
		expect(screen.getByText("Sample Group")).toBeInTheDocument()

		// Retrieves mocked list of group activities
		const techniqueList = screen.getByTestId("technique-exercise-list")

		// Assert if list of techniques is displayed 
		expect(techniqueList).toBeInTheDocument()

		/** 
		 * Below is the way to test whether the inserted json object 
		 * is added to the list of techniques. However, the json object is 
		 * instead inserted into the BeltPicker, and techniques are added 
		 * as filterable belts instead of displayed in the list of techniques 
		 * the group has performed. The json object 'mockedGroupActivities' 
		 * isn't passed to the Statistics component properly. 
		 * No functionality beyond the list of techniques being rendered has been tested.
		*/
		// mockedGroupActivities.forEach(technique => {
		// 	expect(within(techniqueList).getByText(technique.name)).toBeInTheDocument()
		// })
	})

	test("displays error message when data fetching fails", async () => {
		// Suppress console error since test is supposed to generate one
		jest.spyOn(console, "error").mockImplementation(() => {})

		// Mock fetch API to simulate failure
		// eslint-disable-next-line no-undef
		global.fetch = jest.fn(() => Promise.reject(new Error("Failed to fetch")))

		render(<BrowserRouter> <Statistics/> </BrowserRouter>)

		// Wait for error message to be rendered
		await screen.findByText("Gruppen hittades inte")

		// Ensure error message is displayed
		expect(screen.getByText("Gruppen hittades inte")).toBeInTheDocument()
	})

	// Restore the original console error implementation after each test
	afterEach(() => {
		jest.restoreAllMocks()
	})

})


/**
 * Test is incomplete because the BeltPicker cant use the 'mockedBelts' json object.
 * According to its (BeltPickers) documentation its` input should follow the structure
 *  states = [
 *  {
 *   "id": 1,
 *   "name": "Brun",
 *   "color": "FFFFF6",
 *   "child": false
 * 	}
 * ]
 * Which the mockedBelts object does. 
 * Something is preventing the BeltPicker from inserting the data into its list, 
 * and so any functionality beyond being rendered has not been tested.
 */
describe("FilterStatistics component", () => {

	test("FilterStatistics renders correctly within Statistics component", async () => {

		const mockedDates = {
			from: new Date("2022-05-08").toISOString(),
			to: new Date("2024-05-08").toISOString(),
		}

		// eslint-disable-next-line no-undef
		global.fetch = jest.fn(() =>
			Promise.resolve({
				ok: true,
				json: () =>
					Promise.resolve([
						...mockedBelts,
					]),
			})
		)

		// Render statistics page
		render(	
			<BrowserRouter>	
				<FilterStatistics 
					onToggleExercise={jest.fn()}
					onToggleKihon={jest.fn()}
					onDateChanges={jest.fn()}
					onToggleBelts={jest.fn()}
					onClearBelts={jest.fn()}
					belts={mockedBelts}
					dates={mockedDates} /> 
			</BrowserRouter> )

		// Simulate clicking the filter button to open the filter container
		fireEvent.click(screen.getByTestId("filter-button"))

		// Assert existence of filter container
		expect(screen.getByTestId("filter-container")).toBeInTheDocument()

		// Clicking the beltpicker to show dropdown
		// fireEvent.click(screen.getByTestId("techniqueFilter-BeltPicker"))

		// Assert existence of on of the mocked belts 
		// expect(screen.getByText("Svart")).toBeInTheDocument()
	})
})
	



