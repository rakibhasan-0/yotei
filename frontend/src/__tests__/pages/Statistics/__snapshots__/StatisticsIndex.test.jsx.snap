// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics Component Snapshot match 1`] = `
<DocumentFragment>
   
  <div>
    <title>
      Statistik
    </title>
    <svg
      fill="none"
      height="45"
      id=""
      viewBox="0 0 45 45"
      width="45"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M22.5 43C22.5 44.1046 21.6015 45.0093 20.5012 44.9124C9.0111 43.901 0 34.2528 0 22.5C0 10.0736 10.0736 0 22.5 0C34.2528 0 43.901 9.0111 44.9124 20.5012C45.0093 21.6015 44.1046 22.5 43 22.5V22.5C41.8954 22.5 41.0114 21.6014 40.8935 20.5032C39.8977 11.225 32.0426 4 22.5 4C12.2827 4 4 12.2827 4 22.5C4 32.0426 11.225 39.8977 20.5032 40.8935C21.6014 41.0114 22.5 41.8954 22.5 43V43Z"
        fill="#B4B4B4"
        fill-rule="evenodd"
      />
    </svg>
    <div>
      <div>
        <div
          id="filter-container"
        >
          <div>
            <div>
              <button
                class="undefined undefined undefined"
                id="filter-button"
                style="width: 40px;"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.5 2a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM9.05 3a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0V3h9.05zM4.5 7a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM2.05 8a2.5 2.5 0 0 1 4.9 0H16v1H6.95a2.5 2.5 0 0 1-4.9 0H0V8h2.05zm9.45 4a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zm-2.45 1a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0v-1h9.05z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            id="filter-box"
          >
            <fieldset>
              <legend>
                Filtering
              </legend>
              <div>
                <h2>
                  Från
                </h2>
                <div />
                <input
                  class="false undefined"
                  data-testid="date-picker-input"
                  id="start-date-picker"
                  max="6969-04-20"
                  type="date"
                  value="2022-05-16"
                />
              </div>
              <div>
                <h2>
                  Till
                </h2>
                <div />
                <input
                  class="false undefined"
                  data-testid="date-picker-input"
                  id="end-date-picker"
                  max="6969-04-20"
                  type="date"
                  value="2024-05-16"
                />
              </div>
              <div>
                <div
                  id="techniqueFilter-BeltPicker"
                >
                  <div
                    id="techniqueFilter-BeltPicker-header"
                  >
                    <div />
                    <p>
                      Bälten
                    </p>
                    <div
                      class=" "
                      id="techniqueFilter-BeltPicker-toggle-dropdown"
                    >
                      <svg
                        fill="currentColor"
                        height="28"
                        id="techniqueFilter-BeltPicker-dropdown"
                        viewBox="0 0 16 16"
                        width="28"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </div>
                  </div>
                  <div>
                    <div
                      id="techniqueFilter-BeltPicker-children"
                      style="display: none;"
                    >
                      <input
                        style="display: none;"
                        type="checkbox"
                      />
                       
                      <div>
                        <p
                          class="undefined undefined"
                        >
                          Rensa valda bälten
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
                <p />
              </div>
              <p />
              <div>
                <h2>
                  Visa Övningar
                </h2>
                <div />
                <label
                  class="undefined undefined "
                  id="techniqueFilter-VisaÖvningar"
                >
                  <input
                    id="techniqueFilter-VisaÖvningar-checkbox"
                    type="checkbox"
                  />
                </label>
              </div>
              <div>
                <h2>
                  Kihon
                </h2>
                <div />
                <label
                  class="undefined undefined "
                  id="techniqueFilter-KihonCheck"
                >
                  <input
                    id="techniqueFilter-KihonCheck-checkbox"
                    type="checkbox"
                  />
                </label>
              </div>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        id="grading-statistics-container"
      >
        <div>
          <button
            id="popup-button"
          >
            <img
              src="/GradingStatistics.svg"
            />
          </button>
        </div>
      </div>
      <div>
        <button>
          <img
            src="/Statistics.svg"
          />
        </button>
      </div>
    </div>
    <div
      class="activitiesContainer"
      id="technique-exercise-list"
    >
      <h5
        style="font-size: 25px;"
      >
        Inga aktiviteter hittades
      </h5>
    </div>
    <div>
      <button
        class="undefined undefined undefined"
        id="statistics-back-button"
        style="width: 25%;"
        type="button"
      >
        <p>
          Tillbaka
        </p>
      </button>
    </div>
  </div>
   
</DocumentFragment>
`;

exports[`Statistics Filter Snapshot match 1`] = `
<DocumentFragment>
  <div>
    <div
      id="filter-container"
    >
      <div>
        <div>
          <button
            class="undefined undefined undefined"
            id="filter-button"
            style="width: 40px;"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5 2a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM9.05 3a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0V3h9.05zM4.5 7a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM2.05 8a2.5 2.5 0 0 1 4.9 0H16v1H6.95a2.5 2.5 0 0 1-4.9 0H0V8h2.05zm9.45 4a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zm-2.45 1a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0v-1h9.05z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        id="filter-box"
      >
        <fieldset>
          <legend>
            Filtering
          </legend>
          <div>
            <h2>
              Från
            </h2>
            <div />
            <input
              class="false undefined"
              data-testid="date-picker-input"
              id="start-date-picker"
              max="6969-04-20"
              type="date"
              value="2022-05-08T00:00:00.000Z"
            />
          </div>
          <div>
            <h2>
              Till
            </h2>
            <div />
            <input
              class="false undefined"
              data-testid="date-picker-input"
              id="end-date-picker"
              max="6969-04-20"
              type="date"
              value="2024-05-08T00:00:00.000Z"
            />
          </div>
          <div>
            <div
              id="techniqueFilter-BeltPicker"
            >
              <div
                id="techniqueFilter-BeltPicker-header"
              >
                <div />
                <p>
                  Bälten
                </p>
                <div
                  class=" "
                  id="techniqueFilter-BeltPicker-toggle-dropdown"
                >
                  <svg
                    fill="currentColor"
                    height="28"
                    id="techniqueFilter-BeltPicker-dropdown"
                    viewBox="0 0 16 16"
                    width="28"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </div>
              <div>
                <div
                  id="techniqueFilter-BeltPicker-children"
                  style="display: none;"
                >
                  <input
                    style="display: none;"
                    type="checkbox"
                  />
                   
                  <div>
                    <p
                      class="undefined undefined"
                    >
                      Rensa valda bälten
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <p />
          </div>
          <p />
          <div>
            <h2>
              Visa Övningar
            </h2>
            <div />
            <label
              class="undefined undefined "
              id="techniqueFilter-VisaÖvningar"
            >
              <input
                id="techniqueFilter-VisaÖvningar-checkbox"
                type="checkbox"
              />
            </label>
          </div>
          <div>
            <h2>
              Kihon
            </h2>
            <div />
            <label
              class="undefined undefined "
              id="techniqueFilter-KihonCheck"
            >
              <input
                id="techniqueFilter-KihonCheck-checkbox"
                type="checkbox"
              />
            </label>
          </div>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
